iwyu_output.cc

```diff
@@ -1106,11 +1106,27 @@ void ProcessForwardDeclare(OneUse* use,
   }
 
   // (A7) If --no_fwd_decls has been passed, recategorize as a full use unless
-  // the decl is in this file (in which case it must be a self-sufficient decl
-  // being used, so we can just let IWYU do its work).
-  if (!use->ignore_use() &&
-      GlobalFlags().no_fwd_decls &&
-      GetFileEntry(use->decl_loc()) != GetFileEntry(use->use_loc())) {
+  // there is a decl in this file before the use (in which case it must be a
+  // self-sufficient decl being used, so we can just let IWYU do its work).
+  if (!use->ignore_use() && GlobalFlags().no_fwd_decls) {
+    for (const Decl* curr_decl = use->decl(); curr_decl != nullptr;
+         curr_decl = curr_decl->getPreviousDecl()) {
+      if (IsBeforeInSameFile(curr_decl->getLocation(), use->use_loc()))
+        return;
```

Comment:
I'd like to avoid an early return here, because it's not unthinkable that we would like to add additional policies after (A7). Let's see if I can figure out a way to factor that...

```c++
bool promote_to_full_use = true;
for (const Decl* decl = use->decl(); decl != nullptr;
     decl = decl->getPreviousDecl()) {
  // Rationale comment here...
  if (IsBeforeInSameFile(decl->getLocation(), use->use_loc())) {
    promote_to_full_use = false;
    break;
  }

  // Rationale comment here...
  if (IsBeforeInTranslationUnit(decl->getLocation(), use->use_loc())) {
    use->reset_decl(cast<NamedDecl>(decl));
    use->set_suggested_header(
        ConvertToQuotedInclude(GetFilePath(decl)));
    break;
  }
}

if (promote_to_full_use) {
  use->set_full_use();
}
```

I think you can rationalize each of the cases in detail in the loop, and make the `// (A7)` comment a little more high-level.

Reply:
In the changes I did the comments in a different way. I think it is easier to explain what's happening this way. Let me know if you are satisfied with it.